library/psa_crypto.c

```diff
@@ -1166,16 +1166,79 @@ static psa_algorithm_t psa_key_policy_algorithm_intersection(
     /* Common case: both sides actually specify the same policy. */
     if( alg1 == alg2 )
         return( alg1 );
-    /* If the policies are from the same hash-and-sign family, check
-     * if one is a wildcard. If so the other has the specific algorithm. */
-    if( PSA_ALG_IS_HASH_AND_SIGN( alg1 ) &&
-        PSA_ALG_IS_HASH_AND_SIGN( alg2 ) &&
-        ( alg1 & ~PSA_ALG_HASH_MASK ) == ( alg2 & ~PSA_ALG_HASH_MASK ) )
-    {
-        if( PSA_ALG_SIGN_GET_HASH( alg1 ) == PSA_ALG_ANY_HASH )
-            return( alg2 );
-        if( PSA_ALG_SIGN_GET_HASH( alg2 ) == PSA_ALG_ANY_HASH )
-            return( alg1 );
+    /* Special case: one or both sides contain a wildcard. */
+    if( PSA_ALG_IS_WILDCARD( alg1 ) || PSA_ALG_IS_WILDCARD( alg2 ) )
+    {
+        /* If the policies are from the same hash-and-sign family, check
+         * if one is a wildcard. If so the other has the specific algorithm. */
+        if( PSA_ALG_IS_HASH_AND_SIGN( alg1 ) &&
+            PSA_ALG_IS_HASH_AND_SIGN( alg2 ) &&
+            ( alg1 & ~PSA_ALG_HASH_MASK ) == ( alg2 & ~PSA_ALG_HASH_MASK ) )
+        {
+            if( PSA_ALG_SIGN_GET_HASH( alg1 ) == PSA_ALG_ANY_HASH )
+                return( alg2 );
+            if( PSA_ALG_SIGN_GET_HASH( alg2 ) == PSA_ALG_ANY_HASH )
+                return( alg1 );
+        }
+        /* If the policies are from the same AEAD family, then at least one
+         * of them is a minimum-tag-length policy. Calculate the most
+         * restrictive tag length. */
+        if( PSA_ALG_IS_AEAD( alg1 ) && PSA_ALG_IS_AEAD( alg2 ) &&
+            ( PSA_ALG_AEAD_WITH_TAG_LENGTH( alg1, 0 ) ==
+              PSA_ALG_AEAD_WITH_TAG_LENGTH( alg2, 0 ) ) )
+        {
+            size_t alg1_len = PSA_ALG_AEAD_GET_TAG_LENGTH( alg1 );
+            size_t alg2_len = PSA_ALG_AEAD_GET_TAG_LENGTH( alg2 );
+            size_t max_len = alg1_len > alg2_len ? alg1_len : alg2_len;
+
+            /* If both are wildcards, return most restricitve wildcard */
+            if( PSA_ALG_IS_WILDCARD( alg1 ) && PSA_ALG_IS_WILDCARD( alg2 ) )
```

Comment:
I'm not convinced it causes code size increase - I'd rather argue it's the opposite. Both statements have already been evaluated at line 1170, and the input is a local variable, meaning any good compiler will keep track of the evaluation result instead of evaluating the statement again.

And since @athoelke said he didn't want to add more PSA_ALG_IS_xxx macro's, this is the cleanest code for the time being. I'm not convinced there will be other forms of wildcards for this family, and if those do get defined in the future, we can revisit the logic at that time.

Reply:
An implementation can add any number of internal support macros that are consistent with the encoded values that it supports. The spec only wants to define values in the API that have a clear benefit to an application that uses the API.

Without evidence that the compiler can remove the entire enclosing `IS_WILDCARD()` tests from generated code, I am on Gilles' page here. I would not generally accept the following control flow code:

```
if (x == a || x == b || x == c) {
    if (x == a) {
        // handle a case
    } else if (x == b) {
        // handle b case
    } else if (x == c) {
        // handle c case
    }
}
```

The initial if statement is redundant, but a compiler _might_ be able to spot this and remove it. However, if the three conditions themselves become more complex, such as `PSA_ALG_IS_AEAD(x) && (x & PSA_ALG_AEAD_MINIMUM_LENGTH_FLAG) != 0`, then it become increasingly possible that the redundancy is missed and the runtime code pays in space and time for this.

Reply:
Realised that this specific comment applies to the if statement on line 1170, rather than here.

At this point, an optimising bulid _might_ carry the constraint from line 1186 into the complex `IS_WILDCARD()` test here. However, that constraint is expressed as a complex bit mask and test - and so it depends on how the compiler represents established constraints on variable values within code blocks, and whether it could therefore determine that the `IS_HASH_AND_SIGN()` and `IS_MAC()` branches of the `IS_WILDCARD()` evaluation are false. It is more likely that it can detect that the `IS_AEAD()` condition is true, but that evaluation comes after testing for signature and mac algorithm categories.

Reply:
Updated.